File: src/main/java/cspfj/constraint/semantic/Eq.java
package cspfj.constraint.semantic;

import cspfj.constraint.AbstractPVRConstraint;
import cspfj.problem.Domain;
import cspfj.problem.Variable;

public final class Eq extends AbstractPVRConstraint {

	private final int[][] corresponding;

	private final int a, b;

	/**
	 * Constraint x = y.
	 * 
	 * @param x
	 * @param y
	 */
	public Eq(final Variable x, final Variable y) {
		this(1, x, 0, y);
	}

	/**
	 * Constraint ax + b = y.
	 * 
	 * @param a
	 * @param x
	 * @param b
	 * @param y
	 */
	public Eq(final int a, final Variable x, final int b, final Variable y) {
		super(x, y);
		if (a == 0) {
			throw new IllegalArgumentException("a must be != 0");
		}
		final Domain xDom = x.getDomain();
		final Domain yDom = y.getDomain();
		corresponding = new int[][] { new int[xDom.maxSize()],
				new int[yDom.maxSize()] };

		for (int i = xDom.firstIndex(); i >= 0; i = xDom.next(i)) {
			corresponding[0][i] = yDom.index(a * xDom.value(i) + b);
		}

		for (int i = yDom.firstIndex(); i >= 0; i = yDom.next(i)) {
			final int r = yDom.value(i) - b;
			if (r % a == 0) {
				corresponding[1][i] = xDom.index(r / a);
			} else {
				corresponding[1][i] = -1;
			}
		}
		this.a = a;
		this.b = b;
	}

	@Override
	public boolean check() {
		return a * getValue(0) + b == getValue(1);
	}

	public boolean revise(final int position) {
		boolean change = false;
		final Variable var = getVariable(position);
		final Variable otherVar = getVariable(1 - position);
		final int[] correspond = corresponding[position];
		for (int i = var.getFirst(); i >= 0; i = var.getNext(i)) {
			final int index = correspond[i];
			if (index < 0 || !otherVar.isPresent(index)) {
				var.remove(i);
				change = true;
			}
		}
		return change;
	}

	@Override
	public boolean isConsistent(final int reviseCount) {
		final Domain dom = getVariable(0).getDomain();
		final Domain otherDom = getVariable(1).getDomain();
		final int[] correspond = corresponding[0];
		for (int i = dom.firstIndex(); i >= 0; i = dom.next(i)) {
			final int index = correspond[i];
			if (index >= 0 && otherDom.present(index)) {
				return true;
			}
		}
		return false;
	}

	public String toString() {
		final StringBuilder stb = new StringBuilder();
		if (a != 1) {
			stb.append(a).append('.');
		}
		stb.append(getVariable(0));
		if (b > 0) {
			stb.append(" + ").append(b);
		} else if (b < 0) {
			stb.append(" - ").append(-b);
		}
		return stb.append(" = ").append(getVariable(1)).toString();
	}

	@Override
	public float getEvaluation() {
		return getVariable(0).getDomainSize() + getVariable(1).getDomainSize();
	}
}

File: src/main/scala/cspfj/constraint/semantic/Eq.scala
package cspfj.constraint.semantic;

import cspfj.constraint.VariablePerVariable
import cspfj.Domain
import cspfj.Variable
import cspfj.constraint.Constraint
import cspfj.util.Interval
import cspfj.constraint.Shaver

/**
 * Constraint (-)x + b = y.
 *
 * @param a
 * @param x
 * @param b
 * @param y
 */
final class Eq(val neg: Boolean, val x: Variable, val b: Int, val y: Variable)
  extends Constraint(Array(x, y)) with Shaver {

  //  val corresponding = Array(
  //    x.dom.allValues map { v => y.dom.index(a * v + b) },
  //    y.dom.allValues map { v =>
  //      val r = v - b
  //      if (r % a == 0) x.dom.index(r / a) else -1
  //    })

  /**
   * public
   * Constraint x = y.
   *
   * @param x
   * @param y
   */
  def this(x: Variable, y: Variable) = this(false, x, 0, y);

  def checkValues(t: Array[Int]) = (if (neg) -t(0) else t(0)) + b == t(1);

  def yIndex(xIndex: Int) =
    if (neg) y.dom.index(-x.dom.value(xIndex) + b)
    else y.dom.index(x.dom.value(xIndex) + b)

  def xIndex(yIndex: Int) = {
    val yv = y.dom.value(yIndex)
    if (neg) x.dom.index(b - yv) else x.dom.index(yv - b)
  }

  def shave(): List[Int] = {
    var mod: List[Int] = Nil
    if (neg) {
      if (scope(0).dom.intersectVal(scope(1).dom.valueInterval.negate + b)) {
        mod ::= 0
      }
      if (scope(1).dom.intersectVal(scope(0).dom.valueInterval.negate + b)) {
        mod ::= 1
      }
    } else {
      if (scope(0).dom.intersectVal(scope(1).dom.valueInterval - b)) {
        mod ::= 0
      }
      if (scope(1).dom.intersectVal(scope(0).dom.valueInterval + b)) {
        mod ::= 1
      }
    }
    mod
  }

  def reviseVariable(position: Int, mod: List[Int]) = position match {
    case 0 => x.dom.filter { i =>
      val index = yIndex(i)
      index >= 0 && y.dom.present(index)
    }

    case 1 => y.dom.filter { i =>
      val index = xIndex(i)
      index >= 0 && x.dom.present(index)
    }

    case _ => throw new IllegalArgumentException
  }

  override def isConsistent() = {
    val otherDom = y.dom

    x.dom.indices.exists { i =>
      val index = yIndex(i)
      index >= 0 && otherDom.present(index)
    }
  }

  override def toString = (if (neg) "-" else "") + x +
    (if (b > 0) " + " + b else if (b < 0) " - " + (-b) else "") +
    " == " + y

  def getEvaluation = if (x.dom.bound && y.dom.bound) 3 else (x.dom.size + y.dom.size)
  val simpleEvaluation = 2
}
